### DISCUSSION TITLE
Appending query params to existing ones

### DISCUSSION THREAD

**Discussion by kwiat1990** [remix-run/remix#4404] [Q&A] [answered]
Hi, I'm wondering if there is an easy way to handle multiple query parameter within a `<Link>` component? My use case is following: there is an page with filters - each filter appends a query parameter with its label, e.g. `?category=bears`. Additionally on the there's a pagination, so for each filter its results can be also paginated, which gives following URL `?category=bears&page=2`.

For pagination I use `<Link>` component with a `state` props (see: https://github.com/remix-run/remix/issues/186#issuecomment-1031778216) to prevent the whole page from scrolling back to the top and a regex to change the current page:

```ts
const [searchParams, setSearchParams] = useSearchParams();
const pageParam = searchParams.get("page");
const page = pageParam ? +pageParam : 1;

const newParams = searchParams
  .toString()
  .replace(/page=[0-9]/g, `page=${page ? +page + 1 : 2}`);

...

<Link state={{scroll: false}} to={?`${newParams}`}>Load more</Link>
```

I don't like this solution though and I don't think it's really bullet proof in case of even more query parameters. I have tried to use `setSearchParams` to append page with new value inside `useEffect` but of course it gives me an infinite loop as it changes current page and reacts to that change:

```ts
React.useEffect(() => {
  setSearchParams({
      ...Object.fromEntries(searchParams.entries()),
      page: `${page + 1}`,
  });
}, [page]); 

...

<Link state={{scroll: false}} to={?`${searchParams}`}>Load more</Link>
```

Does anyone has an idea how it could be solved?

Edit: I have managed to simplify the code to this:

```ts
const [searchParams] = useSearchParams();
const pageParam = searchParams.get("page") ?? "1";
searchParams.set("page", `${+pageParam + 1}`);

...

<Link state={{scroll: false}} to={{ search: searchParams.toString() }}>Load more</Link>
```

**Comment by kiliman** [comment, marked as answer]
This is how I do it. I created a helper for scoped params. Scopes allows you to manage search params across routes. 

https://github.com/kiliman/remix-scoped-params

**Comment by kwiat1990** [comment]
Thanks for the reply. I went with the last approach and it does work pretty well in my use case.

**Comment by xHomu** [comment]
Try this pattern:
```jsx
 <Link
                to={getFilterOptions(filter)}
>filter</Link>
 
  const location = useLocation();


  const getFilterOptions = (filter: string) => {
    const searchParams = new URLSearchParams(location.search);

    searchParams.set("filter", filter);

    return `?${searchParams.toString()}`;
  };```
